src/wasm/wasm-binary.cpp

```diff
@@ -2126,33 +2160,64 @@ void WasmBinaryBuilder::readNames(size_t payloadLen) {
     auto subsectionPos = pos;
-    if (nameType != BinaryConsts::UserSections::Subsection::NameFunction) {
-      // TODO: locals
-      std::cerr << "warning: unknown name subsection at " << pos << std::endl;
-      pos = subsectionPos + subsectionSize;
-      continue;
-    }
-    auto num = getU32LEB();
-    std::set<Name> usedNames;
-    for (size_t i = 0; i < num; i++) {
-      auto index = getU32LEB();
-      auto rawName = getInlineString();
-      rawName = escape(rawName);
-      auto name = rawName;
-      // De-duplicate names by appending .1, .2, etc.
-      for (int i = 1; !usedNames.insert(name).second; ++i) {
-        name = rawName.str + std::string(".") + std::to_string(i);
+    if (nameType == BinaryConsts::UserSections::Subsection::NameModule) {
+      wasm.name = getInlineString();
+    } else if (nameType ==
+               BinaryConsts::UserSections::Subsection::NameFunction) {
+      auto num = getU32LEB();
+      std::set<Name> usedNames;
+      for (size_t i = 0; i < num; i++) {
+        auto index = getU32LEB();
+        auto rawName = getInlineString();
+        rawName = escape(rawName);
+        auto name = rawName;
+        // De-duplicate names by appending .1, .2, etc.
+        for (int i = 1; !usedNames.insert(name).second; ++i) {
+          name = rawName.str + std::string(".") + std::to_string(i);
+        }
+        // note: we silently ignore errors here, as name section errors
+        //       are not fatal. should we warn?
+        auto numFunctionImports = functionImports.size();
+        if (index < numFunctionImports) {
+          functionImports[index]->name = name;
+        } else if (index - numFunctionImports < functions.size()) {
+          functions[index - numFunctionImports]->name = name;
+        } else {
+          throwError("function index out of bounds: " + std::string(name.str) +
+                     "@" + std::to_string(index));
+        }
       }
-      // note: we silently ignore errors here, as name section errors
-      //       are not fatal. should we warn?
+    } else if (nameType == BinaryConsts::UserSections::Subsection::NameLocal) {
+      auto numFuncs = getU32LEB();
       auto numFunctionImports = functionImports.size();
-      if (index < numFunctionImports) {
-        functionImports[index]->name = name;
-      } else if (index - numFunctionImports < functions.size()) {
-        functions[index - numFunctionImports]->name = name;
-      } else {
-        throwError("index out of bounds: " + std::string(name.str));
+      for (size_t i = 0; i < numFuncs; i++) {
+        auto funcIndex = getU32LEB();
+        Function* func;
+        if (funcIndex < numFunctionImports) {
+          func = functionImports[funcIndex];
+        } else if (funcIndex - numFunctionImports < functions.size()) {
+          func = functions[funcIndex - numFunctionImports];
+        } else {
+          throwError("function index out of bounds: " +
+                     std::to_string(funcIndex));
```

Comment:
It looks like errors in the function subsection are silently ignored, and it would probably make sense to do the same with errors in this subsection. Turning errors in both sections into warnings would also be fine. The important thing is that errors are handled consistently and non-fatally.

Reply:
Sounds good, made all the recoverable errors warnings.